pyplanet/apps/contrib/rankings/__init__.py

```diff
@@ -0,0 +1,243 @@
+import logging
+import math
+
+from pyplanet.apps.contrib.rankings.models.ranked_map import RankedMap
+from pyplanet.apps.contrib.rankings.models import Rank
+from pyplanet.apps.contrib.rankings.views import TopRanksView, MapListView
+from pyplanet.apps.config import AppConfig
+
+from peewee import RawQuery
+
+from pyplanet.apps.core.maniaplanet.models import Player, Map
+from pyplanet.apps.core.maniaplanet import callbacks as mp_signals
+from pyplanet.contrib.command import Command
+from pyplanet.contrib.setting import Setting
+
+logger = logging.getLogger(__name__)
+
+
+class Rankings(AppConfig):
+	# Rankings can only be calculated on Trackmania games.
+	game_dependencies = ['trackmania', 'trackmania_next']
+
+	# Rankings depend on the local records.
+	app_dependencies = ['core.maniaplanet', 'core.trackmania', 'local_records']
+
+	def __init__(self, *args, **kwargs):
+		super().__init__(*args, **kwargs)
+
+		self.setting_records_required = Setting(
+			'minimum_records_required', 'Minimum records to acquire ranking', Setting.CAT_BEHAVIOUR, type=int,
+			description='Minimum of records required to acquire a rank (minimum 3 records).',
+			default=5
+		)
+
+		self.setting_chat_announce = Setting(
+			'rank_chat_announce', 'Display server ranks on map start', Setting.CAT_BEHAVIOUR, type=bool,
+			description='Whether to display the server rank on every map start.',
+			default=True
+		)
+
+		self.setting_topranks_limit = Setting(
+			'topranks_limit', 'Maximum rank to display in topranks', Setting.CAT_BEHAVIOUR, type=int,
+			description='Amount of ranks to display in the topranks view.',
+			default=100
+		)
+
+	async def on_start(self):
+		# Listen to signals.
+		self.context.signals.listen(mp_signals.map.map_end, self.map_end)
+		self.context.signals.listen(mp_signals.player.player_connect, self.player_connect)
+
+		# Register commands.
+		await self.instance.command_manager.register(
+			Command('rank', target=self.chat_rank, description='Displays your current server rank.'),
+			Command('nextrank', target=self.chat_nextrank, description='Displays the player ahead of you in the server ranking.'),
+			Command('topranks', target=self.chat_topranks, description='Displays a list of top ranked players.'),
+		)
+
+		# Register settings
+		await self.context.setting.register(self.setting_records_required, self.setting_chat_announce, self.setting_topranks_limit)
+
+	async def map_end(self, map):
+		# Calculate server ranks.
+		await self.calculate_server_ranks()
+
+		# Display the server rank for all players on the server after calculation, if enabled.
+		chat_announce = await self.setting_chat_announce.get_value()
+		if chat_announce:
+			for player in self.instance.player_manager.online:
+				await self.chat_rank(player)
+
+	async def player_connect(self, player, is_spectator, source, signal):
+		await self.chat_rank(player)
```

Comment:
Shouldn't you check the setting first? Or is this only to announce this privately?